Add support for Ethereum EIP 1559

FrederikBolding opened the pull request:
Fixes #1604 

The following PR adds support for typed Ethereum transactions (EIP 2718) with a focus on EIP 1559. This changes the format of transactions to include the transaction type, new gas parameters as well as an optional access list.

It also adds a new UI screen to show the new gas parameters.
<img width="373" alt="Screenshot 2021-06-06 at 21 34 39" src="https://user-images.githubusercontent.com/1561200/120937671-0c50f080-c70f-11eb-8140-5bcb00799336.png">

I have attempted to re-use as much of the existing transaction signing code as well as added support in the Python lib.

I have repurposed some of the existing tests to cover most of the same cases with the new format. On top of that the PR can be (and has been) tested on the Calaveras testnet (as well as more networks coming later this month).

Resources
https://eips.ethereum.org/EIPS/eip-2718
https://eips.ethereum.org/EIPS/eip-1559
https://eips.ethereum.org/EIPS/eip-2930
https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/client-integration-testnets/calaveras.md

FrederikBolding commented:
This PR looks to be in a working state after a bit of back and worth. I unfortunately hadn't read up on your commit history policy before starting the work, so feel free to let me know if you want anything changed. 

I look forward to starting the review process and hopefully we can get this merged before EIP 1559 goes live on the Ethereum mainnet.

matejcik commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -22,8 +25,31 @@
+        return await sign_eip1559_tx(ctx, msg, keychain)
+
+    # Remove empty access list for legacy tx
+    if len(msg.access_list) == 0:
```
what happens if access_list is not empty for legacy tx?

FrederikBolding replied:
This is sort of an edge case with `protobuf`. Because if you pass `access_list = None` in the `EthereumSignTx` message, it turns it into an empty array. But for legacy transactions we expect the `access_list = None`, if not, the RLP length is wrong and thereby also the signature.

FrederikBolding replied:
Reading this again, I probably should just set `access_list = None` before entering legacy signing anyway. In case the user has passed something the signature will be invalid.

FrederikBolding replied:
I have simplified the logic in `get_total_length` to consider this instead, which imo is cleaner. Thanks for making me aware!

matejcik replied:
> But for legacy transactions we expect the `access_list = None`, if not, the RLP length is wrong and thereby also the signature.

no idea where you got this: `sign_legacy_tx` doesn't touch the `access_list` field so it should not care about whether there's an empty list or anything else

matejcik replied:
> I have simplified the logic in `get_total_length` to consider this instead, which imo is cleaner. Thanks for making me aware!

oh, i see now! yes, that looks cleaner :)

FrederikBolding commented:
@matejcik Any clue why the `defs_check` is failing? I don't see difference in my message to other messages.

matejcik commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -22,8 +25,27 @@
 @with_keychain_from_chain_id
 async def sign_tx(ctx, msg, keychain):
+
+    if msg.tx_type == 2:
```
does it make sense to set `msg.tx_type == 2` and at the same time `msg.eip2718` is False? what happens then?

FrederikBolding replied:
No, it doesn't. My latest commit adds a check for that.

FrederikBolding replied:
Feel free to let me know if you would prefer something else.

FrederikBolding replied:
The `check` function will throw if the `eip2718_type` is out of range. So I think my latest commit should be fine for this.

FrederikBolding replied:
But let me know if that isn't the case.

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -101,19 +127,115 @@ async def sign_tx(ctx, msg, keychain):
+    data_total = msg.data_length
+
+    # detect ERC - 20 token
+    token = None
```
Done!

FrederikBolding commented:
@matejcik How would you prefer I handle the conflicts? Should I wait until you finish the review to do rebasing etc?

matejcik commented:
> Should I wait until you finish the review to do rebasing etc?

yes, please do not rebase or force-push while the review is ongoing.
refer to https://docs.trezor.io/trezor-firmware/misc/review.html

FrederikBolding commented:
> > Should I wait until you finish the review to do rebasing etc?
> 
> yes, please do not rebase or force-push while the review is ongoing.
> refer to https://docs.trezor.io/trezor-firmware/misc/review.html

Yeah, that's what I thought. Just wanted to confirm. Let me know when you want me to handle the conflicts then.

matejcik commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -101,19 +128,93 @@ async def sign_tx(ctx, msg, keychain):
 
     length += rlp.field_length(len(msg.nonce), msg.nonce[:1])
-    length += rlp.field_length(len(msg.gas_price), msg.gas_price)
+    if not msg.eip2718 and msg.gas_price is not None:
```
all this should be under a single `if msg.eip2718_type is not None:`

also all the fields should be mandatory
you can change the protobuf definition to `required` for all fields that are used both in legacy and new-style transactions.
you are already checking the `max_priority_fee` in `check()`, so you don't need to repeat `is not None` for everything here

FrederikBolding replied:
I've tried to simplify this in my latest commit. `access_list` should be present on other types of EIP2718 transactions though, so I can't group it all under one if statement.

matejcik commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -101,19 +128,93 @@ async def sign_tx(ctx, msg, keychain):
 
-    if msg.chain_id:  # forks replay protection
+    if msg.eip2718:
+        length += rlp.field_length(1, [msg.chain_id])
```
don't you need to figure out the length in the same way as below?

FrederikBolding replied:
Yeah, good catch. I just had issues with extra length you guys were adding for a reason not entirely clear to me 😄

matejcik commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -183,9 +288,10 @@ def check(msg: EthereumSignTx):
 
 def check_gas(msg: EthereumSignTx) -> bool:
-    if msg.gas_price is None or msg.gas_limit is None:
+    isEIP1559 = msg.eip2718 and msg.tx_type == 2
```
`is_eip1559`

FrederikBolding replied:
Done

matejcik reviewed (changes requested):
I finally did a detailed review and proposed some fundamental changes to the protobuf. If you agree that the changes make sense, please implement them.

Now would be a good time to rebase, but you might run into a problem with the layout function (see comment), so I leave it up to you if you want to implement that now, or if you want to hold off rebasing until we implement the `confirm_amount` function on our side.

matejcik commented:
in any case, thanks for working on this so far!

FrederikBolding commented on core/src/apps/ethereum/layout.py:
```diff
@@ -40,6 +40,23 @@ async def require_confirm_fee(
     await require_hold_to_confirm(ctx, text, ButtonRequestType.SignTx)
 
 
+async def require_confirm_eip1559_fee(
```
I'll hold off on rebasing until that is in then, can you let me know when it is?

FrederikBolding replied:
Any clue as to when that function will be available?

matejcik replied:
i'm working on related things at the moment, presumably end of the week when this is ready in a PR. maybe a week or two more until master

FrederikBolding replied:
Okay, would it be required to use it before merging? The mainnet release is approaching and this might not be ready if blocked for multiple weeks 🤔

matejcik replied:
we are not planning an August release at the moment anyway so presumably this will go out in September

FrederikBolding replied:
That is unfortunate, would definitely prefer that this was live with mainnet 😄

FrederikBolding replied:
With my rebase I have moved the layouts code to the same place as the other legacy UIs. If you still want me to use the new function I can rebase again when that is in.

FrederikBolding replied:
I have now rebased and used this function. See latest commit!

FrederikBolding commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,16 @@ message EthereumSignTx {
-    optional uint32 tx_type = 10;           // (only for Wanchain)
+    optional uint32 tx_type = 10;           // (Used for Wanchain and EIP2718)
+    optional bytes max_gas_fee = 12;        // <=256 bit unsigned big endian (in wei) - EIP 1559
+    optional bytes max_priority_fee = 13;   // <=256 bit unsigned big endian (in wei) - EIP 1559
```
I don't know if I agree with this, I think it might make it more confusing to use the legacy field for this 🤔

matejcik replied:
You're probably right.

A different thing worth considering, however. Maybe it would be nicer to create a new message type for either:
(a) EIP1559 transaction -- so the eip2718 type would not be included and it would be inferred from the message, or
(b) EIP2718 transaction -- so eip2718 type would need to be specified. The question is, how different do we expect different eip2718 types to be from each other?

The main reason for doing this would be type safety: we could set most fields to `required`, both for the legacy tx and the new tx. This would simplify the situation both for callers (it's obvious which data is needed and when), and for Trezor (easier to verify data correctness).

On Trezor side, very little additional work is needed from the current state, just some reorg and registering the message handler.
On trezorlib side, we would probably have separate calls in `trezorlib.ethereum`, and maybe the same `ethereum sign-tx` in `trezorlib.cli.ethereum`? So that "IF using ethereum and not a fork, AND the user doesn't explicitly specify one or the other, THEN we default to EIP1559"?

I am not sure if this doesn't make things difficult elsewhere though. What do you think?

FrederikBolding replied:
I think this might be overcomplicating things at this point. I expect, in the future, that new transaction types will use the EIP 1559 gas params but with another transaction type number and possibly a new param or two appended. 

I also think that the caller itself should decide the values that go into the transaction instead of the library trying to guess which transaction type you are trying to use.

matejcik replied:
> I also think that the caller itself should decide the values that go into the transaction instead of the library trying to guess which transaction type you are trying to use.

for library usage, yes, absolutely. what I mean above is that the _CLI command_ should make this guess for the user, unless params are explicitly specified

matejcik replied:
> I think this might be overcomplicating things at this point

I wouldn't consider it more complicated than the current implementation. There is very little code duplication between the legacy and EIP1559 signing, and what is there can be very cleanly extracted into a shared function (token handling; transaction data handling).

If in the future another type is specified that _only adds_ fields to current EIP1559 transaction, it's easy to extend `EthereumSignEIP1559Tx` with those fields, and possibly introduce a `tx_type` field if necessary (i'd infer it from the presence or absence of the new fields).

OTOH, if in the future another type is specified to _add and remove_ fields, we'll need to install more `if`s into `check()` and `get_total_length()` and we can't implement strong type checking for the fields that are there (which will allow us to, e.g., get completely rid of `sanitize()` function)

It currently seems to me that a separate `EthereumSignEIP1559Tx` operation is a cleaner solution for possible future extensions. I'm not strongly convinced, but I'd like to see at least some arguments against, if you have them :)

FrederikBolding replied:
It just seems to me that adding a `EthereumSignEIP1559Tx` would result in the API being trickier to use and split into a more advanced API from the callers perspective. If I am thinking of how we wanted to implement it in MyCrypto we wouldn't want to call another function to do EIP 1559, we would rather pass the correct arguments to a single function, that then signs legacy or EIP 1559 or whatever else.

But if you prefer adding the `EthereumSignEIP1559Tx` we can go that route.

FrederikBolding replied:
If that's what you would rather want, I might need a better explanation of how you imagine it would work. @matejcik

FrederikBolding replied:
Because how I imagine it, it would complicate the API here: 
https://github.com/trezor/connect/blob/f71a8e256bf8a51c9bf835291e4969d994d56aff/src/js/core/methods/EthereumSignTransaction.js#L19

matejcik replied:
do you already have code in MyCrypto that can send either legacy or eip1559 transaction? (for some other signing method than Trezor)
can you point me to it?

FrederikBolding replied:
We use a common abstraction for most wallets which can be found here: https://github.com/MyCryptoHQ/wallets

Each integration implements the shared interface https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/wallet.ts

The idea is that they take a shared `TransactionRequest` type which may or may not be EIP 1559 (or something else) and signs them. The app itself would determine the parameters of the transaction and whether it is a EIP 1559 tx or legacy.

Examples:
Private Keys: https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/implementations/non-deterministic/private-key.ts#L11-L14

Ledger (WIP): https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/implementations/deterministic/ledger.ts#L28-L54

Trezor: https://github.com/MyCryptoHQ/wallets/blob/0d55575565dcfc0770fa667d44c9439cae684f38/src/implementations/deterministic/trezor.ts#L27-L53

Adding two separate functions/messages for signing transactions would complicate it in our implementation at least. It doesn't make it impossible or anything, just is less nice to work with from my POV.

matejcik replied:
This could very well be hidden in Trezor Connect, if it would make sense.

actually looking at [TransactionRequest](https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-provider/src.ts/index.ts#L18-L35) it might make sense to accept _that_ in Connect. Is this "ethers.js" an ecosystem-wide thing?

I'll talk to Connect folks too, see what they think.

FrederikBolding replied:
Maybe? I haven't looked to much into the Connect implementation, so I'm not a 100% sure of the ramifications of doing the two message types. It just looks to me as it might complicate things, hence my worries as described above.

`ethers.js` is currently what we (at MyCrypto) prefer, there are other alternatives, so I wouldn't say it is necessarily an ecosystem default, but it is used by many. I don't know if it makes sense for you guys to use it as the request type, that would be up to you.

matejcik replied:
looking at ethers.js [signing code](https://github.com/ethers-io/ethers.js/blob/71b7547f10229f50d8b701611c5e6041b8ed966b/packages/transactions/src.ts/index.ts), it seems to come down to that at a certain point, _some_ piece of code must say:
```
if eip2718_type is None:
	check_legacy_fields(...)
	sign_legacy_tx(...)
elif eip2718_type == X:
	check_X_fields(...)
	sign_X_tx(...)
elif eip2718_type == Y:
	check_Y_fields(...)
	sign_Y_tx(...)
# etc.
```
I would prefer that this was not explicitly coded in Trezor firmware, as the "different message" is a good mechanism that does the above implicitly. Perhaps doing this in Connect would be OK.

I'll let you know once we discuss this with Connect folks.

matejcik replied:
We discussed this with Connect and decided that on the firmware side, it is better for us to split into two messages. The Connect API will most likely be unified, the way you like to use it.

Please create a new message `EthereumSignTxEIP1559` with the appropriate fields, no `tx_type` (we will use type 2 as constant for now).

You will need to add an entry to `messages.proto` to register the new message type (pick a new id in the ~450 range), create `ethereum/sign_tx_eip1559.py`, and register it in `workflow_handlers.py`.

Other than that it should just be a matter of rearranging the code that already exists. Let me know if you encounter any problems.

FrederikBolding replied:
@matejcik Done! Let me know if there is any more small things otherwise I would assume this is good for a final review 😄

FrederikBolding commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,16 @@ message EthereumSignTx {
     optional uint32 data_length = 8;        // Length of transaction payload
     optional uint32 chain_id = 9;           // Chain Id for EIP 155
-    optional uint32 tx_type = 10;           // (only for Wanchain)
+    optional uint32 tx_type = 10;           // (Used for Wanchain and EIP2718)
```
👍

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -101,19 +128,93 @@ async def sign_tx(ctx, msg, keychain):
+
+    sha = HashWriter(sha3_256(keccak=True))
+
+    if msg.tx_type is not None:
```
The assertion wasn't needed and has been removed 😄

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -22,22 +25,37 @@
 MAX_CHAIN_ID = 2147483629
 
 
+def format_access_list(access_list):
```
In my opinion the complexity of writing these functions over just formatting the data correctly and letting the RLP code handle it is not worth it. But I can simplify the `format_access_list` function.

matejcik replied:
You're trading RAM for complexity, and it isn't a good tradeoff for Trezor environment in this case. If the encoding was more complex, maybe, but in this case it's literally just writing number of items first, and that's done in `sign_*` itself anyway. OTOH your approach creates a list of lists (which allocates a bunch of items on top of what's already allocated from protobuf), only to throw it away after encoding.

FrederikBolding replied:
I think you might be underestimating the complexity. But no matter, I have pushed changes now that calculate the length and do the encoding without turning it into a list of lists. I would say it sacrifices readability and ease of understanding, but of course up to you!

matejcik replied:
I see what you mean, but this is mainly because the RLP API is bad. I did a refactor in #1704:

```
def access_list_item_length(item: EthereumAccessList) -> int:
	address_length = rlp.length(bytes_from_address(item.address))
	keys_length = rlp.length(item.storage_keys)
	return rlp.header_length(address_length + keys_length) + address_length + keys_length

def access_list_length(access_list: list[EthereumAccessList]) -> int:
	payload_length = sum(access_list_item_length(i) for i in access_list)
	return rlp.header_length(payload_length) + payload_length

def write_access_list(w: Writer, access_list: list[EthereumAccessList]) -> None:
	payload_length = sum(access_list_item_length(i) for i in access_list)
	rlp.write_header(w, payload_length, rlp.LIST_HEADER_BYTE)
	for item in access_list:
		address_bytes = bytes_from_address(item.address)
		address_length = rlp.length(address_bytes)
		keys_length = rlp.length(item.storage_keys)
		rlp.write_header(w, address_length + keys_length, rlp.LIST_HEADER_BYTE)
		rlp.write(w, address_bytes)
		rlp.write(w, item.storage_keys)
```
the main drawback of this code is that we need to call `bytes_from_address` repeatedly, but I don't see a good way around that

FrederikBolding replied:
I can rebase and use that when it is merged 👍

androolloyd replied:
I had a look through this and let me be the first to say I am not well versed in this ecosystem, but would it be possible to pre process the bytes_from_address information once for each time in the access list as a means of creation of its list entry, so its available on the object itself. 

Would be curious what over head is carried through the workflow in ram vs the cost of processing it each time throughout.

matejcik replied:
> so its available on the object itself

We can't add new properties to protobuf messages unfortunately (they're not regular python objects under the hood).
We _could_ create a list of results of bytes_from_address, so that `address_bytes[i] == bytes_from_address(access_list[i].address)`, but that seems like a needless complication for the amount of computation this will entail in practice.

FrederikBolding replied:
In the latest rebase I use your code. I haven't had a chance to test it as my dev environment is somehow broken. See my other comment.

FrederikBolding replied:
This works in latest! 😄

FrederikBolding commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,16 @@ message EthereumSignTx {
+
+    message EthereumAccessList {
+        required string address = 1;
+        repeated string storage_keys = 2;
```
The protobuf already uses `string` for the `to` address, so I assumed that to be fine?

It also seems simpler to be able to construct access_list like so:
```
access_list=[
                    [
                        "0xde0b295669a9fd93d5f28d9ec85e40f4cb697bae",
                        [
                            "0x0000000000000000000000000000000000000000000000000000000000000003",
                            "0x0000000000000000000000000000000000000000000000000000000000000007",
                        ],
                    ]
             ]
```
Dont mind the formatting here 😆

matejcik replied:
We always use strings for _addresses_, because on most coins the address<->bytes conversion is not straightforward.

OTOH we use bytes for pubkeys in most cases because the conversion usually _is_ straightforward, a public key is just bytes.

For construction, `protobuf.dict_to_proto` will handle the hex conversion by itself, so i'm not sure what the issue is...?

FrederikBolding replied:
So are you just asking for the `storage_keys` to be bytes?

matejcik replied:
yes

FrederikBolding replied:
Done!

FrederikBolding commented:
> in any case, thanks for working on this so far!

No worries! Thanks for reviewing!

I did a few commits to simplify and fix some of the issues you pointed out. For a few of your comments I suggested a different approach or disagreed with your suggestion.

Looking forward to your responses on my comments and a re-review.

Will also rebase when you guys have the layout stuff merged, let me know when that is.

matejcik commented:
just a note, do not mark my comments resolved, instead comment with "fixed" or commit hash in which the comment is resolved.

only the person who started the comment should resolve it

FrederikBolding commented:
> just a note, do not mark my comments resolved, instead comment with "fixed" or commit hash in which the comment is resolved.
> 
> only the person who started the comment should resolve it

Sure thing! My bad.

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -101,19 +128,93 @@ async def sign_tx(ctx, msg, keychain):
+
+    data_total = msg.data_length
+
+    await require_confirm_tx(ctx, recipient, value, msg.chain_id, token, msg.tx_type)
```
Done

matejcik reviewed (commented):
A couple more comments, not a complete review round.
I just came back from my vacation so I'll be responding more quickly now :)

FrederikBolding commented on python/src/trezorlib/cli/ethereum.py:
```diff
@@ -192,6 +213,10 @@ def sign_tx(
     token,
+    max_gas_fee,
+    max_priority_fee,
+    access_list=None,
```
Yeah, I don't know specifically how to pass arrays and complex objects via the CLI & `click`. Any ideas?

matejcik replied:
You can pass arrays by using `multiple=True` on the option and passing several of the same option. However, the access list is an array of objects, which would need some sort of custom processing.

The simplest encoding would be something like `-a 0xaddress:01FF02EE:03FF03AA:...`

An alternative is to pass a JSON string (or file) encoding the data, and then use `protobuf.from_dict` to decode it to what we need.

I'll leave up to you which you think is best or easiest. I don't expect this CLI option to be used that frequently and I don't see any particular usecase to optimize for.

matejcik replied:
see `_amount_to_int` usage in options for a relatively nice way to streamline the custom processing

FrederikBolding replied:
I'll try to do something. I think you are correct in assuming the `access_list` parameter will be only used in very few cases.

FrederikBolding replied:
Done!

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -161,9 +257,18 @@ def sign_digest(msg: EthereumSignTx, keychain, digest):
     if msg.tx_type not in [1, 6, None]:
         raise wire.DataError("tx_type out of bounds")
 
+    if msg.tx_type == 2 and (msg.max_gas_fee is None or msg.max_priority_fee is None):
```
Done

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -22,22 +25,36 @@
 
-    data_total = msg.data_length
+    if msg.eip2718_type == 2:
+        return await sign_eip1559_tx(ctx, msg, keychain)
```
Done

FrederikBolding commented on tests/device_tests/test_msg_ethereum_signtx_eip1559.py:
```diff
@@ -0,0 +1,385 @@
+
+@pytest.mark.altcoin
+@pytest.mark.ethereum
+class TestMsgEthereumSigntx:
```
I don't quite understand what you are asking me to do here for the tests? I have constructed the test cases myself based on local testing.

FrederikBolding replied:
Done!

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -39,28 +39,48 @@ def get_public_node(client, n, show_display=False):
     )
 
 
+def format_access_list(access_list):
```
Agreed!

FrederikBolding commented on python/src/trezorlib/cli/ethereum.py:
```diff
@@ -166,8 +177,18 @@ def get_public_node(client, address, show_display):
+@click.option("-x", "--tx-type", type=int, help="TX type")
 @click.option("-t", "--token", help="ERC20 token address")
+@click.option(
+    "-mg", "--max-gas-fee", help="Max Gas Fee (EIP1559)", callback=_amount_to_int
```
Done

FrederikBolding commented on python/src/trezorlib/cli/ethereum.py:
```diff
@@ -192,6 +213,10 @@ def sign_tx(
+    max_gas_fee,
+    max_priority_fee,
+    access_list=None,
+    eip2718_type=None,
```
Done

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -39,28 +39,48 @@ def get_public_node(client, n, show_display=False):
     tx_type=None,
+    max_gas_fee=None,
+    max_priority_fee=None,
+    access_list=None,
```
Done

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -39,28 +39,48 @@ def get_public_node(client, n, show_display=False):
     gas_limit,
     to,
     value,
+    gas_price=None,
```
Done

FrederikBolding commented:
> A couple more comments, not a complete review round.
> I just came back from my vacation so I'll be responding more quickly now :)

Hope you had a nice vacation, welcome back! 😄  I looked through your comments and addressed them all today. Good for another round of review and perhaps some discussion in a few of the comments!

matejcik commented on tests/device_tests/test_msg_ethereum_signtx_eip1559.py:
```diff
@@ -0,0 +1,385 @@
+            eip2718_type=2,
+        )
+
+    # taken from T1 might not be 100% correct but still better than nothing
```
this comment is probably incorrect :) dtto below

FrederikBolding replied:
Yep - fixed!

matejcik commented on python/src/trezorlib/cli/ethereum.py:
```diff
@@ -74,6 +75,25 @@ def _amount_to_int(ctx, param, value):
+        raise click.BadParameter("Access List format invalid")
+
+
+def _parse_access_list_item(ctx, param, value):
```
this probably doesn't need the `ctx, param` arguments?

FrederikBolding replied:
Nope, fixed!

FrederikBolding commented:
@matejcik I have rebased today and moved the UI to the new file and included the new access_list code. This has somehow broken my dev environment. I get `ImportError: no module named 'trezor.messages.Failure'` when trying to run the emulator now 🤔  I have ran `nix-shell` which updated and `poetry install`.

matejcik commented:
did you rebuild the emulator binary? protobufs are now compiled in

FrederikBolding commented:
> did you rebuild the emulator binary? protobufs are now compiled in

Yeah. The binary looks to be compiling fine, it's running the emu.py that is failing.

matejcik commented:
maybe you have something like `from trezor.messages.Failure import Failure` forgotten somewhere? that's the old-style import, with new-style you import `from trezor.message import Failure` directly

FrederikBolding commented:
> maybe you have something like `from trezor.messages.Failure import Failure` forgotten somewhere? that's the old-style import, with new-style you import `from trezor.message import Failure` directly

Iirc it was this line failing: https://github.com/trezor/trezor-firmware/blob/13b8a6e54d3ffd46143551763c364337864701f2/core/src/trezor/wire/__init__.py#L41

I'm not at a computer right now. Can you run my branch locally?

matejcik commented:
works fine for me from your branch. it really looks like your emulator binary is miscompiled. make sure you run `make gen` from root, and try `make clean` in core before rebuilding the emulator

FrederikBolding commented:
> works fine for me from your branch. it really looks like your emulator binary is miscompiled. make sure you run `make gen` from root, and try `make clean` in core before rebuilding the emulator

I tried all of that just now with no luck.

I get this error 
```
[nix-shell:~/dev/trezor-firmware/core]$ poetry run ./emu.py
TREZOR_PATH=udp:127.0.0.1:21324
TREZOR_PROFILE_DIR=/private/var/tmp
TREZOR_UDP_PORT=21324
TREZOR_FIDO2_UDP_PORT=21326
TREZOR_SRC=/Users/frederik/dev/trezor-firmware/core/src
Waiting for emulator to come up...
Traceback (most recent call last):
  File "main.py", line 44, in <module>
  File "boot.py", line 4, in <module>
  File "trezor/wire/__init__.py", line 41, in <module>
ImportError: no module named 'trezor.messages.Failure'
Traceback (most recent call last):
  File "/Users/frederik/dev/trezor-firmware/core/./emu.py", line 286, in <module>
    cli()
```

matejcik commented:
try the following, see what fails:
```sh
[trezor-firmware/core]$ ./build/unix/trezor-emu-core -i                                                                                                                  
MicroPython 57b1188be on 2021-07-20; trezor-emulator version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import trezorproto
>>> type(trezorproto)
<class 'module'>
>>> from trezor import messages
>>> type(messages)
<class 'module'>
>>> type(messages.Failure)
<class 'MsgDef'>
```

FrederikBolding commented:
> try the following, see what fails:
> 
> ```shell
> [trezor-firmware/core]$ ./build/unix/trezor-emu-core -i                                                                                                                  
> MicroPython 57b1188be on 2021-07-20; trezor-emulator version
> Use Ctrl-D to exit, Ctrl-E for paste mode
> >>> import trezorproto
> >>> type(trezorproto)
> <class 'module'>
> >>> from trezor import messages
> >>> type(messages)
> <class 'module'>
> >>> type(messages.Failure)
> <class 'MsgDef'>
> ```

```
[nix-shell:~/dev/trezor-firmware/core]$ ./build/unix/trezor-emu-core -i
MicroPython fd56d3ef7 on 2021-07-20; trezor-emulator version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import trezorproto
>>> type(trezorproto)
<class 'module'>
>>> from trezor import messages
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'trezor'
```

🤔

matejcik commented:
that last one is OK for non-frozen build.
how about this? 
```
[trezor-firmware/core]$ cd src
[trezor-firmware/core/src]$ ../build/unix/trezor-emu-core -i                                                                                                                  
MicroPython 57b1188be on 2021-07-20; trezor-emulator version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import trezorproto
>>> failure = trezorproto.type_for_name("Failure")
>>> type(failure)
<class 'MsgDef'>
>>> from trezor import messages
>>> type(messages)
<class 'module'>
>>> type(messages.Failure)
<class 'MsgDef'>
```

FrederikBolding commented:
> that last one is OK for non-frozen build.
> how about this?
> 
> ```
> [trezor-firmware/core]$ cd src
> [trezor-firmware/core/src]$ ../build/unix/trezor-emu-core -i                                                                                                                  
> MicroPython 57b1188be on 2021-07-20; trezor-emulator version
> Use Ctrl-D to exit, Ctrl-E for paste mode
> >>> import trezorproto
> >>> failure = trezorproto.type_for_name("Failure")
> >>> type(failure)
> <class 'MsgDef'>
> >>> from trezor import messages
> >>> type(messages)
> <class 'module'>
> >>> type(messages.Failure)
> <class 'MsgDef'>
> ```

```
[nix-shell:~/dev/trezor-firmware/core]$ cd src

[nix-shell:~/dev/trezor-firmware/core/src]$ ../build/unix/trezor-emu-core -i
MicroPython fd56d3ef7 on 2021-07-20; trezor-emulator version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import trezorproto
>>> failure = trezorproto.type_for_name("Failure")
>>> type(failure)
<class 'MsgDef'>
>>> from trezor import messages
>>> type(messages)
<class 'module'>
>>> type(messages.Failure)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'Failure'
```

matejcik commented:
is it possible that you still have `src/trezor/messages` directory? It was changed to a file `messages.py`, but presumably didn't go away if there were modified files.

FrederikBolding commented:
> is it possible that you still have `src/trezor/messages` directory? It was changed to a file `messages.py`, but presumably didn't go away if there were modified files.

Yep, that was it. My bad 🤦‍♂️

FrederikBolding commented:
@matejcik With the rebase I have fixed the access_list RLP comments and moved the UI layout to the same place as the legacy components. In my optics the only blocking things are now if you want to wait for the new layout function and the possibility of using another message type for EIP 1559 transactions. I have commented on both these open discussions.

matejcik commented:
Correct, I believe this is mostly done. I might have some minor style comments in the future, based on how we resolve the "another message type" thing (which i'm continuing in the respective thread).

wrt layout, please wait until the updated layout functions are in from my side

(jsyk i also deleted the part where we resolve your build problem, except for the resolution itself)

FrederikBolding commented:
@matejcik What's the latest?

matejcik commented:
`confirm_amount` is now in master so you can use that

re the other issue, i haven't had the opportunity to properly discuss with Connect team yet

FrederikBolding commented:
> `confirm_amount` is now in master so you can use that

Done!

> re the other issue, i haven't had the opportunity to properly discuss with Connect team yet

Otherwise it might be worth it to merge this and you can do the refactor at a later point? 🤔

matejcik commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,26 @@ message EthereumSignTx {
+
+message EthereumSignTxEIP1559 {
+    repeated uint32 address_n = 1;          // BIP-32 path to derive the key from master node
+    optional bytes nonce = 2;               // <=256 bit unsigned big endian
```
most of those fields should be `required`

FrederikBolding replied:
Done

matejcik commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,26 @@ message EthereumSignTx {
+    optional bytes gas_limit = 5;           // <=256 bit unsigned big endian
+    optional string to = 11;                // recipient address
+    optional bytes value = 6;               // <=256 bit unsigned big endian (in wei)
+    optional bytes data_initial_chunk = 7;  // The initial data chunk (<= 1024 bytes)
```
please use: `optional bytes data_initial_chunk = 7 [default=''];`

FrederikBolding replied:
Done

matejcik commented on core/src/apps/ethereum/sign_tx_eip1559.py:
```diff
@@ -0,0 +1,166 @@
+from trezor.utils import HashWriter
+
+from apps.common import paths
+from apps.ethereum.sign_tx import (
```
please use relative import: `from .sign_tx`

FrederikBolding replied:
Done

matejcik commented on core/src/apps/ethereum/sign_tx_eip1559.py:
```diff
@@ -0,0 +1,166 @@
+        raise wire.DataError("missing required params for EIP1559")
+
+    if msg.chain_id < 0:
+        raise wire.DataError("chain_id out of bounds")
```
you can remove all the above checks when the fields are set to `required`.

FrederikBolding replied:
Done

matejcik reviewed (changes requested):
A bunch of small things, yes.

This is not yet a full review, i'll have a more thorough look on Monday. But all in all I believe we're basically there :)

matejcik commented on python/src/trezorlib/ethereum.py:
```diff
@@ -83,6 +89,49 @@ def sign_tx(
+        data, chunk = data[data_length:], data[:data_length]
+        response = client.call(messages.EthereumTxAck(data_chunk=chunk))
+
+    return response.signature_v, response.signature_r, response.signature_s
```
Given that you don't use `chain_id` in the signature, I think you will need to do the appropriate modification here. (per EIP-155 the sig_v must be `sig_byte + chain_id * 2 + 35`)

Or does EIP1559 invalidate that?

FrederikBolding replied:
`chain_id` is part of the EIP 1559 tx payload and is therefore not packed in `v` anymore. This is intended.

matejcik replied:
i see! it's not even `sig_v` anymore. thanks.

matejcik commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,26 @@ message EthereumSignTx {
+    required bytes max_gas_fee = 3;        // <=256 bit unsigned big endian (in wei)
+    required bytes max_priority_fee = 4;   // <=256 bit unsigned big endian (in wei)
+    required bytes gas_limit = 5;           // <=256 bit unsigned big endian
+    optional string to = 11;                // recipient address
```
also please renumber the fields as they should be :)

(the reason EthereumSignTx is not numbered sequentially is because of an incompatible change in the past)

FrederikBolding replied:
Done

matejcik reviewed (changes requested):
Fínal (hopefully) round of comments for the core code.

One thing i'm wondering about is the `chain_id` calculation in `sign_digest`, which you removed. I don't think we actually need it, as it can always be always done on the client. But still wondering.

Also maybe worth considering setting `chain_id` to `bytes` in protobuf right now? (see #1741)

matejcik commented:
I pushed f60c48c to your branch that fixes build on legacy.

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -83,6 +89,49 @@ def sign_tx(
+        access_list=access_list,
+    )
+
+    if data:
```
Done

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -83,6 +89,49 @@ def sign_tx(
+    data=None,
+    chain_id=None,
+    max_gas_fee=None,
+    max_priority_fee=None,
```
Done

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -83,6 +89,49 @@ def sign_tx(
+    gas_limit,
+    to,
+    value,
+    data=None,
```
Done

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -83,6 +89,49 @@ def sign_tx(
+def sign_tx_eip1559(
+    client,
+    n,
+    nonce,
```
Done

FrederikBolding commented on python/src/trezorlib/ethereum.py:
```diff
@@ -51,16 +51,22 @@ def sign_tx(
     data=None,
     chain_id=None,
     tx_type=None,
+    max_gas_fee=None,
```
Done

FrederikBolding commented on core/src/apps/ethereum/sign_tx.py:
```diff
@@ -23,29 +23,17 @@
 async def sign_tx(ctx, msg, keychain):
     msg = sanitize(msg)
+
+    if msg.tx_type not in [1, 6, None]:
```
Not intentional at this point, fixed.

FrederikBolding commented on common/protob/messages-ethereum.proto:
```diff
@@ -66,7 +66,26 @@ message EthereumSignTx {
+    optional string to = 11;                // recipient address
+    required bytes value = 6;               // <=256 bit unsigned big endian (in wei)
+    optional bytes data_initial_chunk = 7 [default=''];  // The initial data chunk (<= 1024 bytes)
+    optional uint32 data_length = 8;        // Length of transaction payload
```
Done

FrederikBolding commented:
> Fínal (hopefully) round of comments for the core code.
🤞 

> One thing i'm wondering about is the `chain_id` calculation in `sign_digest`, which you removed. I don't think we actually need it, as it can always be always done on the client. But still wondering.
As explained in another comment, not needed for EIP 1559 transactions!

> Also maybe worth considering setting `chain_id` to `bytes` in protobuf right now? (see #1741)
I would prefer not to at this point, because it might add extra complexity and require even more back and forth, reviews, etc. If that is okay with you 😄

matejcik reviewed (approved):
squashed & rebased as 204d31ef6d03aa2e71aa9ac97294a9c07632b817
pipeline running at https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/350729690

i'll still need to add a changelog entry

FrederikBolding commented:
> squashed & rebased as [204d31e](https://github.com/trezor/trezor-firmware/commit/204d31ef6d03aa2e71aa9ac97294a9c07632b817)
> pipeline running at https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/350729690
> 
> i'll still need to add a changelog entry

Cool! What needs to be done for trezor-connect? Just regen the protobuf and add params for EIP1559? Is that something you guys are doing or not?

matejcik commented:
> Is that something you guys are doing or not?

It's on the roadmap but nobody is working on it currently. PR will be appreciated :)

I am not sure what's needed honestly. There's some tomfoolery about old versions of protobuf in Connect, so YMMV.

FrederikBolding commented:
> > Is that something you guys are doing or not?
> 
> It's on the roadmap but nobody is working on it currently. PR will be appreciated :)
> 
> I am not sure what's needed honestly. There's some tomfoolery about old versions of protobuf in Connect, so YMMV.

Okay 🤔 

It looks like all tests passed except for the changelog thing and something related to the new message?

matejcik commented:
should be fixed now as a39870449ffae07b53080a3a9e758ebb0e48a374, pipeline at https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/350749141
mind if I force-push this into your branch? that way we'll get the results inline

FrederikBolding commented:
> should be fixed now as [a398704](https://github.com/trezor/trezor-firmware/commit/a39870449ffae07b53080a3a9e758ebb0e48a374), pipeline at https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/350749141
> mind if I force-push this into your branch? that way we'll get the results inline

Go ahead 😄

FrederikBolding commented:
All green! 😄

matejcik commented:
not automatically but I just pushed it out

matejcik commented:
so this should be all on firmware side!

wonderful. thanks again for your contribution and work on this

FrederikBolding commented:
> so this should be all on firmware side!
> 
> wonderful. thanks again for your contribution and work on this

Indeed! Thanks so much for all the reviews 😄

uj commented:
Issue found:
https://github.com/trezor/trezor-firmware/issues/1817
